File: genie/public/js/impersonation.js
// Copyright (c) 2023, Wahni IT Solutions Pvt. Ltd. and Contributors
// MIT License. See license.txt


frappe.ui.form.on("User", {
	refresh: function(frm) {
		if (!frm.doc.__islocal && frm.doc.enabled && frappe.user.has_role("System Manager")) {
			frm.add_custom_button(__("Impersonate"), function() {
				frappe.call({
					method: "genie.utils.impersonation.generate_impersonation_url",
					args: {
						"user": frm.doc.name
					},
					callback: function(r) {
						if (!r.exc && r.message) {
							let impersonate_url = `<a href=${r.message}>this link</a>`;
							frappe.msgprint(
								__("Please open {0} in an incognito window to impersonate {1}", [impersonate_url, frm.doc.name]),
							);
						}
					}
				});
			}, __("Password"));
		}
	}
});

File: genie/utils/impersonation.py

# Copyright (c) 2023, Wahni IT Solutions Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.www.login import _generate_temporary_login_link


@frappe.whitelist()
def generate_impersonation_url(user):
    frappe.only_for("System Manager")

    if not frappe.db.get_single_value("Genie Settings", "enable_user_impersonation"):
        frappe.throw(_("User Impersonation is disabled"))
    if user == "Administrator":
        frappe.throw(_("You cannot impersonate Administrator"))

    return _generate_temporary_login_link(user, 1)